### PR TITLE
Introduce Asset Upload Service

### PR DESCRIPTION
**Pull request by pkosiec**
**Description**

Changes proposed in this pull request:

- Introduce Asset Upload Service

**Related issue(s)**
See #2653

### CODE REVIEW

File: docs/asset-store/docs/assets/asset-upload-service-swagger.yaml [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+                items:
+                  type: string
+    uploadMultiStatus:
+      description: Multi status response. Some files have been uploaded successfully, some didn't
+      content:
+        application/json:
+          schema:
+            type: object
+            properties:
+              errors:
```
Comment (michal-hudy) [inline docs/asset-store/docs/assets/asset-upload-service-swagger.yaml]: It should also provide filename that failed.

File: CODEOWNERS [hunk 2]
```diff
@@ -39,9 +39,12 @@
 # The `configurations-generator` subchart
 /resources/core/charts/configurations-generator @piotrmsc @kubadz @sjanota @strekm @jakkab @Tomasz-Smelcerz-SAP @kfurgol @Demonsthere
 
-# The `assetstore` subchart
+# The `assetstore` chart
 /resources/assetstore/ @akucharska @michal-hudy @pkosiec @mjasinski5 @aerfio @tgorgol @magicmatatjahu
 
+# The `asset-upload-service` chart
+/resources/assetstore/charts/asset-upload-service/ @akucharska @michal-hudy @pkosiec @mjasinski5 @aerfio @tgorgol @magicmatatjahu
```
[the remark this answers is not in the document]
Comment (pkosiec) [reply CODEOWNERS]: Erm... right

File: components/asset-upload-service/internal/configurer/configurer.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+		if apierrors.IsNotFound(err) {
+			return nil, false, nil
+		}
+		return nil, false, errors.Wrapf(err, "while loading ConfigMap %s from %s", c.cfg.Name, c.cfg.Namespace)
+	}
+
+	sharedAppConfig := c.fromConfigMap(configMap)
+	glog.Infof("Config successfully loaded from ConfigMap %s in namespace %s", c.cfg.Name, c.cfg.Namespace)
+
+	return sharedAppConfig, true, nil
```
[the remark this answers is not in the document]
Comment (pkosiec) [reply components/asset-upload-service/internal/configurer/configurer.go]: 😄 right

File: components/asset-upload-service/internal/requesthandler/requesthandler.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+	}
+	r.writeResponse(w, status, Response{
+		UploadedFiles: uploadedFiles,
+		Errors:        errMessages,
+	})
+}
+
+func (r *RequestHandler) generateDirectoryName() string {
+	unixTime := time.Now().Unix()
+	return strconv.FormatInt(unixTime, 32)
```
[the remark this answers is not in the document]
Comment (pkosiec) [reply components/asset-upload-service/internal/requesthandler/requesthandler.go]: Do you mean using an external library? I think that's not necessary for now as our solution should work well 🙂 If there will be some problems, we can migrate to some library like go-uuid.

File: components/asset-upload-service/internal/bucket/handler.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+func (h *Handler) tryCreatingBucket(prefix string) (string, error) {
+	bucketName := h.generateBucketName(prefix)
+
+	for i := 0; i < creationRetries; i++ {
+		glog.Infof("Trying to create a bucket with prefix %s (attempt %d of %d)", prefix, i+1, creationRetries)
+		err := h.CreateIfDoesntExist(bucketName, h.cfg.Region)
+		if err != nil {
+			if i == creationRetries-1 {
```
[the remark this answers is not in the document]
Comment (pkosiec) [reply components/asset-upload-service/internal/bucket/handler.go]: Hmm, I thought about that, but retrying on every error is not bad I think

File: docs/asset-store/docs/assets/asset-upload-service-swagger.yaml [hunk 6]
```diff
@@ -0,0 +1,153 @@
+openapi: 3.0.0
+info:
+  description: The purpose of this application is to provide an API to enable temporary
```
Comment (kazydek) [inline docs/asset-store/docs/assets/asset-upload-service-swagger.yaml]:
```suggestion
  description: The purpose of this application is to provide an API to enable a temporary
```

File: docs/asset-store/docs/assets/asset-upload-service-swagger.yaml [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    host files for Asset Store, without need to use external providers.
```
Comment (kazydek) [inline docs/asset-store/docs/assets/asset-upload-service-swagger.yaml]:
```suggestion
   files hosting service for the Asset Store, without the need to rely on external providers.
```

File: docs/asset-store/docs/assets/asset-upload-service-swagger.yaml [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+  title: Asset Upload Service API
+  license:
+    name: Apache 2.0
+    url: https://www.apache.org/licenses/LICENSE-2.0
+  version: 0.0.1
+paths:
+  /upload:
+    post:
+      description: Uploads files sent with form to Minio.
```
Comment (kazydek) [inline docs/asset-store/docs/assets/asset-upload-service-swagger.yaml]:
```suggestion
      description: Uploads files to Minio as multipart forms.
```

File: docs/asset-store/docs/assets/asset-upload-service-swagger.yaml [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+      summary: Uploads files to Minio.
+      operationId: uploadReq
+      requestBody:
+        content:
+          multipart/form-data:
+            schema:
+              type: object
+              properties:
+                public:
+                  description: Files, which should be uploaded to public system bucket.
```
Comment (kazydek) [inline docs/asset-store/docs/assets/asset-upload-service-swagger.yaml]:
```suggestion
                  description: Files to upload to a public system bucket.
```

File: docs/asset-store/docs/assets/asset-upload-service-swagger.yaml [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+                  type: array
+                  items:
+                    type: string
+                    format: binary
+                private:
+                  description: Files, which should be uploaded to private system bucket.
```
Comment (kazydek) [inline docs/asset-store/docs/assets/asset-upload-service-swagger.yaml]:
```suggestion
                  description: Files to upload to a private system bucket.
```

File: docs/asset-store/docs/assets/asset-upload-service-swagger.yaml [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+                  type: array
+                  items:
+                    type: string
+                    format: binary
+                directory:
+                  description: Directory, where files will be put in bucket. If not specified, the directory name will be generated. If directory and files already exist, they will be overwritten.
```
Comment (kazydek) [inline docs/asset-store/docs/assets/asset-upload-service-swagger.yaml]:
```suggestion
                  description: The directory for storing the uploaded files. If you do not specify it, the service creates a directory with a random name. If the directory and files already exist, the service overwrites them.
```

File: docs/asset-store/docs/assets/asset-upload-service-swagger.yaml [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+          schema:
+            type: object
+            properties:
+              errors:
+                description: Array of error messages
+                type: array
+                items:
+                  $ref: "#/components/schemas/RequestError"
+    uploadMultiStatus:
+      description: Multi status response. Some files have been uploaded successfully, some didn't
```
Comment (kazydek) [inline docs/asset-store/docs/assets/asset-upload-service-swagger.yaml]:
```suggestion
      description: A multi-status response. Only some files have been uploaded successfully.
```

File: docs/asset-store/docs/assets/asset-upload-service-swagger.yaml [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+          schema:
+            type: object
+            properties:
+              errors:
+                description: Array of error messages
+                type: array
+                items:
+                  $ref: "#/components/schemas/FileUploadError"
+              uploadedFiles:
+                description: Files, that were uploaded successfully
```
Comment (kazydek) [inline docs/asset-store/docs/assets/asset-upload-service-swagger.yaml]:
```suggestion
                description: Files that were uploaded successfully.
```

File: docs/asset-store/docs/assets/asset-upload-service-swagger.yaml [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+                  $ref: "#/components/schemas/FileUploadResult"
+    uploadSuccess:
+      description: Success response
+      content:
+        application/json:
+          schema:
+            type: object
+            properties:
+              uploadedFiles:
+                description: Files, that were uploaded successfully
```
Comment (kazydek) [inline docs/asset-store/docs/assets/asset-upload-service-swagger.yaml]:
```suggestion
                description: Files that were uploaded successfully.
```

File: docs/asset-store/docs/assets/asset-upload-service-swagger.yaml [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+      type: object
+      properties:
+        fileName:
+          description: "Name of the file, which upload failed"
+          type: string
+        message:
+          description: "Error message"
+          type: string
+    RequestError:
+      description: "It represents a request failure"
```
Comment (kazydek) [inline docs/asset-store/docs/assets/asset-upload-service-swagger.yaml]:
```suggestion
      description: "It represents a request failure."
```

File: docs/asset-store/docs/assets/asset-upload-service-swagger.yaml [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+        size:
+          description: "File size"
+          type: integer
+          format: int64
+    FileUploadError:
+      description: "It represents a single failure of file upload"
+      type: object
+      properties:
+        fileName:
+          description: "Name of the file, which upload failed"
```
Comment (kazydek) [inline docs/asset-store/docs/assets/asset-upload-service-swagger.yaml]:
```suggestion
          description: "The name of the file which the service failed to upload."
```

File: docs/asset-store/docs/assets/asset-upload-service-swagger.yaml [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+          type: string
+        remotePath:
+          description: "Full remote path to the file. If the bucket is public, it can be accessed with simple GET request"
+          type: string
+        size:
+          description: "File size"
+          type: integer
+          format: int64
+    FileUploadError:
+      description: "It represents a single failure of file upload"
```
Comment (kazydek) [inline docs/asset-store/docs/assets/asset-upload-service-swagger.yaml]:
```suggestion
      description: "It represents a single file upload failure."
```

File: docs/asset-store/docs/assets/asset-upload-service-swagger.yaml [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+            type: object
+            properties:
+              uploadedFiles:
+                description: Files, that were uploaded successfully
+                type: array
+                items:
+                  $ref: "#/components/schemas/FileUploadResult"
+  schemas:
+    FileUploadResult:
+      description: "It represents a single successful file upload result"
```
Comment (kazydek) [inline docs/asset-store/docs/assets/asset-upload-service-swagger.yaml]:
```suggestion
      description: "It represents a single file successful upload result."
```

File: docs/asset-store/docs/assets/asset-upload-service-swagger.yaml [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+      type: object
+      properties:
+        bucket:
+          description: "Bucket name, where the uploaded file is located"
```
Comment (kazydek) [inline docs/asset-store/docs/assets/asset-upload-service-swagger.yaml]:
```suggestion
          description: "The name of the bucket with the uploaded file."
```

File: docs/asset-store/docs/assets/asset-upload-service-swagger.yaml [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+          type: string
+        fileName:
+          description: "File name"
+          type: string
+        remotePath:
+          description: "Full remote path to the file. If the bucket is public, it can be accessed with simple GET request"
```
Comment (kazydek) [inline docs/asset-store/docs/assets/asset-upload-service-swagger.yaml]:
```suggestion
          description: "A full remote path to the file. If the bucket is public, you can access it with a simple GET request."
```

File: docs/asset-store/docs/03-04-asset-upload-service.md [hunk 21]
```diff
@@ -0,0 +1,71 @@
+---
+title: Asset Upload Service
+type: Details
+---
+
+Asset Upload Service is a HTTP server, that exposes file upload functionality for Minio. It contains a simple HTTP endpoint, which accepts multiple files as form. It can upload files for two system buckets: private and public one, with read-only policy set.
```
[the remark this answers is not in the document]
Comment (pkosiec) [reply docs/asset-store/docs/03-04-asset-upload-service.md]: I meant that only the public bucket has the read-only policy set

[the remark this answers is not in the document]
Comment (pkosiec) [reply docs/asset-store/docs/03-04-asset-upload-service.md]: Actually we can omit the read-only policy, as it is described in one of sections below

[the remark this answers is not in the document]
Comment (pkosiec) [reply docs/asset-store/docs/03-04-asset-upload-service.md]: `It contains a simple HTTP endpoint which accepts multiple files as forms`
hmm I would rewrote it

File: components/asset-upload-service/README.md [hunk 22]
```diff
@@ -0,0 +1,153 @@
...
+         "size": 4414
+      }
+   ]
+}
+```
+
+See the [Swagger specification](../../docs/asset-store/docs/assets/asset-upload-service-swagger.yaml) to read full API documentation. You can use the [Swagger Editor](https://editor.swagger.io) to preview and test the API service.
+
+## Configuration
+
```
[the remark this answers is not in the document]
Comment (pkosiec) [reply components/asset-upload-service/README.md]: 😢 I suspected that

File: docs/asset-store/docs/assets/asset-upload-service-swagger.yaml [hunk 23]
```diff
@@ -0,0 +1,152 @@
...
+  description: The purpose of this application is to provide an API to enable files hosting service for the Asset Store, without the need to rely on external providers.
+  title: Asset Upload Service API
+  license:
+    name: Apache 2.0
+    url: https://www.apache.org/licenses/LICENSE-2.0
+  version: 0.0.1
+paths:
+  /upload:
+    post:
+      description: Uploads files to Minio from `multipart/form-data` forms
```
Comment (kazydek) [inline docs/asset-store/docs/assets/asset-upload-service-swagger.yaml]:
```suggestion
      description: Uploads files to Minio from `multipart/form-data` forms.
```

Comment (kazydek) [reply docs/asset-store/docs/assets/asset-upload-service-swagger.yaml]: I'm not sure if that formatting is required.

**Pull request merged**